### ISSUE TITLE
Port twisted.names.client to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6056]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6056|
|Type|enhancement|
|Created|2012-10-04 18:02:56Z|
|Branch|https://github.com/twisted/twisted/tree/dns-client-py3-6056-2|

Because it doesn't run on Python 3 now and the goal is for it to do so.

<details><summary>Searchable metadata</summary>

```
trac-id__6056 6056
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_dns_client_py3_6056_2 branches/dns-client-py3-6056-2
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1349373776000000 1349373776000000
changetime__1349980341000000 1349980341000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [36043]) Branching to 'dns-client-py3-6056'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [36062]) Branching to 'dns-client-py3-6056-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @itamarst|
|-|-|

Porting occurred.  There's also some `twisted.names.root` porting in this branch.  Necessary for the same reason as `hosts` and `cache`, but I realized partway through that while it needs to _import_ for the minimal milestone, it doesn't need to _work_, so I immediately stopped.  Result is a slightly-more-than-necessary amount of porting in `root.py` and its tests.  I left it out of the ported list.

Other stuff is the usual, `zope.interface`, bytes vs unicode, unpacking in parameter definitions, except syntax, etc.  More conversions of `str(Name)` to `Name.name` to preserve bytesness.  I had to add a bunch of missing test coverage for `createResolver` and I also deleted support for the deprecated `Resolver.protocol` attribute (the tests still used this deprecated attribute, so they needed changin').

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/dns-client-py3-6056-2)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

I'll deal with this tomorrow, probably, but a note on `root`: it is possible (likely?) we will initially have `setup.py` installing only modules that are in the ported list. If a module that is required for code to run is not there, this will break things. I opened a ticket for porting `twisted.python.win32` as a result, you should either do the same, or remove the dependency on `root` importing, or start thinking of alternate `setup.py` strategies. In the latter case you should open the ticket for `setup.py` porting to discuss this, since we need to do it real soon now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [36072]) Remove another deprecated, now broken, use of Resolver.protocol


refs [#6056](https://github.com/twisted/twisted/issues/6056)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst set owner to @exarkun|
|-|-|

Hooray for more tests.

  1. `getWarningMethod` is unnecessary import now in `client.py`.
  1. Change news file to be removal, to note removal of deprecated `Resolver.protocol`.
  1. Some failures in the news test on Windows need to be fixed.
  1. In `test_client.py`, a typo: "A missing nameserver configuration file *is* results".
  1. `test_names` has at least two classes that are really testing `client.py` (`FilterAnswerTests`, `RetryLogic`), maybe others as well. Perhaps they should be moved to `test_client.py` for better Python 3 coverage?
  1. `run-python3-tests` doesn't pass either. Possibly you forgot a final commit? E.g. `parseConfig` still uses `file()` in version I'm seeing.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks.  Sorry about the half-completed work.

> 1. getWarningMethod is unnecessary import now in client.py.
> 2. Change news file to be removal, to note removal of deprecated Resolver.protocol.
> 4. In test_client.py, a typo: "A missing nameserver configuration file *is* results".

Fixed in r36080

> 3. Some failures in the news test on Windows need to be fixed.

I punted, instead :/  Testing thread-using code is hard.  I filed [#6098](https://github.com/twisted/twisted/issues/6098) and marked the tests as skipped.

> 5. test_names has at least two classes that are really testing client.py (FilterAnswerTests, RetryLogic), maybe others as well. Perhaps they should be moved to test_client.py for better Python 3 coverage?

Done in r36090 and r36091.

> 6. run-python3-tests doesn't pass either. Possibly you forgot a final commit? E.g. parseConfig still uses file() in version I'm seeing.

Huh, probably got distracted by writing new tests and forgot to re-run them under Python 3.  Fixed in r36087.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst set owner to @exarkun|
|-|-|

Thanks!

It still looks like there's a couple of failed tests on Windows, unless I misread the build; skipping as a temporary solution is fine. Fix, then merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [36094]) Skip more tests on Windows

refs [#6056](https://github.com/twisted/twisted/issues/6056)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [36097]) Merge dns-client-py3-6056-2

Author: exarkun
Reviewer: itamarst
Fixes: [#6056](https://github.com/twisted/twisted/issues/6056)

Port `twisted.names.client` (and `twisted.names.resolve` and some small bits of
`twisted.names.root`) to Python 3.  Also remove the deprecated `Resolver.protocol`
attribute, remove some uses of it from Twisted, and add some missing test coverage.
